### ISSUE TITLE
HttpResponse can't be deserialized when cached in Redis

### ISSUE THREAD

**Issue by hampos** [TomPallister/Ocelot#152]
This feels like a platform specific issue, I'm running Ocelot in a docker container from the `microsoft/aspnetcore:2.0` image which is a Debian implementation, but haven't tested it directly on Windows.
The issue happens when using a `CacheHandler` with Json Serialization and serializing an `HttpResponse`.
It seems that Json.NET can't deserialize the Unix implementation of HttpResponse `CurlResponseMessage` since it doesn't have a valid constructor.

```
Exception caught in global error handler, exception message: Unable to find a constructor to use for type System.Net.Http.CurlHandler+CurlResponseMessage. A class should either have a default constructor, one constructor with arguments or a constructor marked with the JsonConstructor attribute. Path 'Version', line 1, position 11., exception stack:    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)
```

**Comment by hampos** [comment]
@TomPallister 
An update on this, with the `InMemoryJsonHandle` from the pull request https://github.com/TomPallister/Ocelot/pull/151
I believe I created an acceptance test that exposes this.
It seems it isn't a platform specific issue after all, and `HttpResponse` version of Windows implementation can't be deserialized from Json as well.
I have created a pull request with the test on develop
https://github.com/TomPallister/Ocelot/pull/153

**Comment by TomPallister** [comment]
@hampos might need to map it to something else as you suggested!

**Comment by hampos** [comment]
@TomPallister 
I did a little digging in .Net Core's implementation on Response Caching.
They seem to be storing in their cache a `MemoryCachedResponse` object
https://github.com/aspnet/ResponseCaching/blob/f616f3e891405d89ad21f4ff53848ca289784fb6/src/Microsoft.AspNetCore.ResponseCaching/Internal/MemoryResponseCache.cs#L60

which is a transformation of a `CachedResponse`
https://github.com/aspnet/ResponseCaching/blob/f616f3e891405d89ad21f4ff53848ca289784fb6/src/Microsoft.AspNetCore.ResponseCaching/Internal/CacheEntry/CachedResponse.cs#L10

Then in the middleware they map the CachedResponse to the HttpContext's Response.
https://github.com/aspnet/ResponseCaching/blob/f616f3e891405d89ad21f4ff53848ca289784fb6/src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs#L165-L183

**Comment by TomPallister** [comment]
@hampos how is this middleware used? Is it something we can already register in asp.net core and use instead of the output caching I wrote for Ocelot? Or is it something we can just pull in and use instead? Or is this something we can copy to get it working with CacheManager? I can't see that this middleware hooks into a distributed cache for example? But we could add it I guess as its open source?! Lots of options.

Awesome stuff btw! Really appreciate the help.

**Comment by hampos** [comment]
@TomPallister Yeah, my thoughts exactly, I was looking a way to hook it up with a distributed cache and didn't see any.
The middleware exists in the `Microsoft.AspNetCore.ResponseCaching` or the `Microsoft.AspNetCore.All` package. It seems pretty sophisticated, uses `Cache-Control` headers etc.
Will do some research on how to make it work with CacheManager and report back.

**Comment by hampos** [comment]
@TomPallister 
I started doing some work on porting the `ResponseCachingMiddleware` to Ocelot to merge its functionality with `OutputCacheMiddleware`.
The problem is that it uses various stuff declared as internal by other classes in the package. I finally had to move the entire package source inside Ocelot for it to access all its dependencies.
Then I realized this is pointless, we will be porting unmantainable code without being able to pull versions from the original repo.
In the issue tracker there is an issue about supporting distributed caching, I placed a question there about opening the middleware's constructor in order to accept other caching implementations, they said it's not in their immediate plans.
https://github.com/aspnet/ResponseCaching/issues/137

Any ideas on how to tackle this now?
I'm thinking to fork the ResponseCaching repo, opening the constructor and push a new package with that change in Nuget. Then install this in Ocelot and create a DistributedResponseCache implementation using IOcelotCache to inject in it. Then we'll just have to sync the fork and push new packages when they release versions.
The problem again is that this middleware is not compatible with Ocelot's pipeline and settings...
